File: python/src/falcon_core/physics/device_structures/impedances.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.impedances import Impedances as _CImpedances
from falcon_core.physics.device_structures.impedance import Impedance
from falcon_core.generic.list import List

class Impedances:
    """Python wrapper for Impedances."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def new_empty(cls, ) -> Impedances:
        return cls(_CImpedances.new_empty())

    @classmethod
    def new(cls, items: List) -> Impedances:
        obj = cls(_CImpedances.new(items._c if items is not None else None))
        obj._ref_items = items  # Keep reference alive
        return obj

    @classmethod
    def from_json(cls, json: str) -> Impedances:
        return cls(_CImpedances.from_json(json))

    def push_back(self, value: Impedance) -> None:
        ret = self._c.push_back(value._c if value is not None else None)
        return ret

    def size(self, ) -> None:
        ret = self._c.size()
        return ret

    def empty(self, ) -> None:
        ret = self._c.empty()
        return ret

    def erase_at(self, idx: Any) -> None:
        ret = self._c.erase_at(idx)
        return ret

    def clear(self, ) -> None:
        ret = self._c.clear()
        return ret

    def at(self, idx: Any) -> Impedance:
        ret = self._c.at(idx)
        if ret is None: return None
        return Impedance._from_capi(ret)

    def items(self, ) -> List:
        ret = self._c.items()
        if ret is None: return None
        return List(ret)

    def contains(self, value: Impedance) -> None:
        ret = self._c.contains(value._c if value is not None else None)
        return ret

    def intersection(self, other: Impedances) -> Impedances:
        ret = self._c.intersection(other._c if other is not None else None)
        return Impedances._from_capi(ret)

    def index(self, value: Impedance) -> None:
        ret = self._c.index(value._c if value is not None else None)
        return ret

    def equal(self, other: Impedances) -> None:
        ret = self._c.equal(other._c if other is not None else None)
        return ret

    def not_equal(self, other: Impedances) -> None:
        ret = self._c.not_equal(other._c if other is not None else None)
        return ret

    def to_json(self, ) -> str:
        ret = self._c.to_json()
        return ret

    def __len__(self):
        return self.size()

    def __getitem__(self, idx):
        ret = self.at(idx)
        if ret is None:
            raise IndexError("Index out of bounds")
        return ret

    def append(self, value):
        return self.push_back(value)

    @classmethod
    def from_list(cls, items):
        return cls(_CImpedances.from_list(items))

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, Impedances):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, Impedances):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/_opam/.opam-switch/sources/falcon_core/src/falcon_core/physics/device_structures/impedances.ml
open Ctypes
open Capi_bindings
open ErrorHandling

open Falcon_core.Generic
open Falcon_core.Physics.Device_structures

class c_impedances (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.impedances_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module Impedances = struct
  type t = c_impedances

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.impedances_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_impedances ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.impedances_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_impedances ptr

  let empty () : t =
    let ptr = Capi_bindings.impedances_create_empty () in
    ErrorHandling.raise_if_error ();
    new c_impedances ptr

  let make (items : ListImpedance.t) : t =
    ErrorHandling.read items (fun () ->
      let ptr = Capi_bindings.impedances_create items#raw in
      ErrorHandling.raise_if_error ();
      new c_impedances ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.impedances_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.impedances_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.impedances_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let pushBack (handle : t) (value : Impedance.t) : unit =
    ErrorHandling.multi_read [handle; value] (fun () ->
      let result = Capi_bindings.impedances_push_back handle#raw value#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let size (handle : t) : int =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.impedances_size handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let empty (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.impedances_empty handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let eraseAt (handle : t) (idx : int) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.impedances_erase_at handle#raw (Unsigned.Size_t.of_int idx) in
      ErrorHandling.raise_if_error ();
      result
    )

  let clear (handle : t) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.impedances_clear handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let at (handle : t) (idx : int) : Impedance.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.impedances_at handle#raw (Unsigned.Size_t.of_int idx) in
      ErrorHandling.raise_if_error ();
      new c_impedance result
    )

  let items (handle : t) : ListImpedance.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.impedances_items handle#raw in
      ErrorHandling.raise_if_error ();
      new c_listimpedance result
    )

  let contains (handle : t) (value : Impedance.t) : bool =
    ErrorHandling.multi_read [handle; value] (fun () ->
      let result = Capi_bindings.impedances_contains handle#raw value#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let intersection (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.impedances_intersection handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_impedances result
    )

  let index (handle : t) (value : Impedance.t) : int =
    ErrorHandling.multi_read [handle; value] (fun () ->
      let result = Capi_bindings.impedances_index handle#raw value#raw in
      ErrorHandling.raise_if_error ();
      result
    )

end